Fix error of deleting files in afterEach, before asynchronous child p…

lechlm opened the pull request:
…rocess has finished. I just have realized that the test case with the child process never terminates in the CI workflow because afterEach (which deleted the config file) is directly called after spawning the child process (I don't know why this worked on my local machine). I fixed this issue by creating temp folders for every test case in beforeEach. As far as I found out, there is no real other possibility for handling this.

lechlm committed:
Fix error of deleting files in afterEach, before asynchronous child process has finished

alessiogambi reviewed (changes requested):
Please make sure that the tests remove the temp folder after the tests finish. 
It might be also a good idea to enforce a timeout for each test.

alessiogambi commented:
According to the documentation, enforcing the timeout for the javascript tests should not be necessary, since mocha should have already a default timeout of 2sec in place.

lechlm committed:
Implemented removal of temporary folders + refactor existing tests.

lechlm commented:
I implemented the automatic removal of the temporary folder (though it might be bit of a hack, as elaborated in the code comment). Furthermore, I have already refactored the existing tests (according to issue #14) because it was easier than considering both the child processes and the "direct call" of the program. However, without using process.exit, the tests of course will fail from now on.

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
-            var exitCode = minesweeper.main("simple.cfg")
-            assert.strictEqual(exitCode, 1, "Wrong exit code for not existent board configuration file")
+        it('Minesweeper should exit with code 1 if input file does not exist', function () {
+            minesweeperProcess = spawn('node', [minesweeperHome + "minesweeper.js", "simple.cfg"]);
```
Shouldn't this be something like `tempDir + simple.cfg`? The point is: if someone has an existing `simple.cfg` file, the test might fail for the wrong reason. Instead, by "pretending" the file is in the tempDir we should be sure that indeed there will be no file as the test expects.

Question: `tempDir` is defined as `os.tmpdir() + path.sep` on line 44, doesn't this implies that there's a tralining "/" already?

lechlm replied:
It should be indeed tempDir + simple.cfg, yes. Otherwise exactly the issue mentioned might happen.

tempDir is not exactly defined as os.tmpdir() + path.sep, but as a subfolder of it (using fs.mkdtempSync). So the trailing "/" is necessary because mkdtempSync returns the path without this sign. However, it might be a bit error-prone that every time a config file is created, the separator has to be written. So I append the separator immediately in the beforeEach function (cf. recent commit).

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
+        it('Minesweeper should exit with code 2 if input file is invalid (empty)', function () {
             // Make sure there's an empty "simple.cfg" file (see https://flaviocopes.com/how-to-create-empty-file-node/)
-            fs.closeSync(fs.openSync("simple.cfg", 'w'))
+            fs.closeSync(fs.openSync(tempDir + "/simple.cfg", 'w'))
```
See comment above on defining `tempDir` as `os.tmpdir() + path.sep`. I suspec that the `/` might be superfluous/wrong?

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
+            fs.closeSync(fs.openSync(tempDir + "/simple.cfg", 'w'))
 
-            var stats = fs.statSync("simple.cfg")
+            var stats = fs.statSync(tempDir + "/simple.cfg")
```
Same comment as above

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
+            var stats = fs.statSync(tempDir + "/simple.cfg")
 
-            var exitCode = minesweeper.main("simple.cfg")
+            minesweeperProcess = spawn('node', [minesweeperHome + "minesweeper.js", tempDir + "/simple.cfg"]);
```
Same comment as above on `tempDir` and `/`

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
-            var exitCode = minesweeper.main("simple.cfg")
-
-            assert.strictEqual(exitCode, 0, "Wrong exit for valid (won) game")
+            fs.writeFileSync(tempDir + "/simple.cfg", lines.join('\n') + '\n')
```
As above

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
-            assert.strictEqual(exitCode, 0, "Wrong exit for valid (won) game")
+            fs.writeFileSync(tempDir + "/simple.cfg", lines.join('\n') + '\n')
+
+            minesweeperProcess = spawn('node', [minesweeperHome + "minesweeper.js", tempDir + "/simple.cfg"]);
```
As above

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -8,103 +9,137 @@ const { stdin } = require('mock-stdin');
             const lines = ['..*', '...', '...'];
-            fs.writeFileSync("simple.cfg", lines.join('\n') + '\n')
-            var lengthUI = (lines.length * 2 + 1) + 3; // first part for the board + three lines for the message box
+            fs.writeFileSync(tempDir + "/simple.cfg", lines.join('\n') + '\n')
```
As above

alessiogambi commented on javascript/test/basic-tests.js:
```diff
@@ -125,31 +160,27 @@ describe('Basic Tests', function() {
             // spawn child process to execute Minesweeper instance
-            minesweeperProcess = spawn('node', [minesweeperHome + "minesweeper.js", "simple.cfg"]);
-            
+            minesweeperProcess = spawn('node', [minesweeperHome + "minesweeper.js", tempDir + "/simple.cfg"]);
```
As above

alessiogambi reviewed (changes requested):
I have a doubt about the definition of tempDir and its usage. Please address the comments and it should be ready to be merged.

lechlm committed:
Some minor improvements.

lechlm committed:
Use path.sep instead of only '/'.

alessiogambi commented:
I found a few additional issues and implemented some improvement over your pull request. For example, I am experimenting with the use of the sync version of `spawn` to avoid problems due to async behaviors during tests execution (not very javascript-ish but better than mixing async and sync computations)

lechlm commented:
I also thought about using the sync version, but in favor of executing the asserts after every GUI refresh (which in my opinion is benefitial when wanting to check just intermediate moves) I refrained from using it. However, because of the mixing of async and sync computations (and maybe also because of the other issues you found?), using the sync version might be the better approach.

alessiogambi commented:
Yes, interactively checking the system response to each input sound nice. But for the sake of our testing, also prefixing all the inputs, and asserting over the entire output produced works fine.
